File: src/server/services/github.ts
import type {
  Commit,
  PullRequest,
  PullRequestCIStatus,
  PullRequestReviewStatus,
  Review,
  GitHubEvent,
  GitHubRepoCommit,
  GitHubSearchResponse,
  GitHubSearchItem,
  GitHubSearchCommitResponse,
  GitHubSearchCommitItem,
  GitHubUser,
  GitHubPullRequestCommit,
  GitHubPullRequestReview,
  GitHubPullRequestDetails,
  GitHubCombinedStatus,
  GitHubCheckRun,
  GitHubCheckRunsResponse,
  GitHubIssueComment,
  GitHubPullRequestReviewComment,
  PullRequestDiscussionComment,
} from "./github.types";
import { GitHubRateLimitError } from "./github.types";

const GITHUB_API = "https://api.github.com";

// --- Pure mapping functions (exported for testing) ---

export function parseNextUrl(linkHeader: string | null): string | null {
  if (!linkHeader) return null;
  const match = /<([^>]+)>;\s*rel="next"/.exec(linkHeader);
  return match?.[1] ?? null;
}

export function extractRepoName(repositoryUrl: string): string {
  const parts = repositoryUrl.split("/repos/");
  return parts[1] ?? repositoryUrl;
}

export function mapPushEventToCommits(event: GitHubEvent): Commit[] {
  if (event.type !== "PushEvent" || !event.payload.commits) return [];
  const branch = event.payload.ref?.replace("refs/heads/", "") ?? null;
  return event.payload.commits.map((c) => ({
    sha: c.sha,
    message: c.message.split("\n")[0]!,
    repoName: event.repo.name,
    branch,
    url: `https://github.com/${event.repo.name}/commit/${c.sha}`,
    createdAt: new Date(event.created_at),
  }));
}

export function mapSearchItemToPR(item: GitHubSearchItem): PullRequest {
  const repoName = extractRepoName(item.repository_url);
  let state: PullRequest["state"];
  if (item.pull_request?.merged_at) {
    state = "merged";
  } else if (item.state === "open" && item.draft) {
    state = "draft";
  } else if (item.state === "open") {
    state = "open";
  } else {
    state = "closed";
  }
  return {
    title: item.title,
    repoName,
    url: item.html_url,
    state,
    createdAt: new Date(item.created_at),
    updatedAt: new Date(item.updated_at),
  };
}

export function mapSearchItemToReview(item: GitHubSearchItem): Review {
  const repoName = extractRepoName(item.repository_url);
  const state = item.pull_request?.merged_at
    ? "merged"
    : item.state === "open"
      ? "open"
      : "closed";

  return {
    title: item.title,
    repoName,
    url: item.html_url,
    state,
    createdAt: new Date(item.created_at),
    updatedAt: new Date(item.updated_at),
  };
}

export function extractPullRequestNumber(url: string): number | undefined {
  const match = /\/pull\/(\d+)(?:\/|$)/.exec(url);
  const value = match?.[1];
  return value ? Number(value) : undefined;
}

export function mapSearchCommitToCommit(
  item: GitHubSearchCommitItem,
  branch?: string,
): Commit {
  return {
    sha: item.sha,
    message: item.commit.message.split("\n")[0]!,
    repoName: item.repository.full_name,
    branch: branch ?? null,
    url: item.html_url,
    createdAt: new Date(item.commit.author.date),
  };
}

// --- API client ---

async function githubFetch(token: string, url: string): Promise<Response> {
  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: "application/vnd.github.v3+json",
    },
  });

  if (response.status === 403) {
    const remaining = response.headers.get("x-ratelimit-remaining");
    if (remaining === "0") {
      const resetTimestamp = response.headers.get("x-ratelimit-reset");
      const resetAt = new Date(Number(resetTimestamp) * 1000);
      throw new GitHubRateLimitError(resetAt);
    }
  }

  if (!response.ok) {
    throw new Error(
      `GitHub API error: ${response.status} ${response.statusText}`,
    );
  }

  return response;
}

// --- Fetch functions ---

export async function fetchLogin(token: string): Promise<string> {
  const response = await githubFetch(token, `${GITHUB_API}/user`);
  const user = (await response.json()) as GitHubUser;
  return user.login;
}

export async function fetchCommits(
  token: string,
  login: string,
): Promise<Commit[]> {
  const [eventCommits, searchCommits] = await Promise.all([
    fetchCommitsFromEvents(token, login),
    fetchCommitsFromSearch(token, login),
  ]);

  const seen = new Set<string>();
  const merged: Commit[] = [];
  for (const commit of [...eventCommits, ...searchCommits]) {
    if (!seen.has(commit.sha)) {
      seen.add(commit.sha);
      merged.push(commit);
    }
  }

  return merged.toSorted(
    (a, b) => b.createdAt.getTime() - a.createdAt.getTime(),
  );
}

async function fetchCommitsFromEvents(
  token: string,
  login: string,
): Promise<Commit[]> {
  const branches = new Map<string, string>();
  let url: string | null = `${GITHUB_API}/users/${login}/events?per_page=100`;
  let page = 0;

  while (url && page < 3) {
    const response = await githubFetch(token, url);
    const events = (await response.json()) as GitHubEvent[];
    for (const event of events) {
      if (event.type !== "PushEvent" || !event.payload.ref) continue;
      const branch = event.payload.ref.replace("refs/heads/", "");
      const key = `${event.repo.name}:${branch}`;
      if (!branches.has(key)) {
        branches.set(key, `${event.repo.name}:${branch}`);
      }
    }
    url = parseNextUrl(response.headers.get("link"));
    page++;
  }

  const BASE_BRANCHES = new Set(["main", "master", "next", "develop", "dev"]);
  const branchKeys = [...branches.keys()].toSorted((a, b) => {
    const branchA = splitRepoBranch(a)[1];
    const branchB = splitRepoBranch(b)[1];
    const aBase = BASE_BRANCHES.has(branchA) ? 0 : 1;
    const bBase = BASE_BRANCHES.has(branchB) ? 0 : 1;
    return aBase - bBase;
  });

  const since = new Date(Date.now() - 90 * 24 * 60 * 60 * 1000).toISOString();
  const seen = new Set<string>();
  const result: Commit[] = [];

  for (const key of branchKeys) {
    const [repoName, branch] = splitRepoBranch(key);
    try {
      const res = await githubFetch(
        token,
        `${GITHUB_API}/repos/${repoName}/commits?author=${login}&sha=${encodeURIComponent(branch)}&per_page=100&since=${since}`,
      );
      const items = (await res.json()) as GitHubRepoCommit[];
      for (const c of items) {
        if (seen.has(c.sha)) continue;
        seen.add(c.sha);
        result.push({
          sha: c.sha,
          message: c.commit.message.split("\n")[0]!,
          repoName,
          branch,
          url: c.html_url,
          createdAt: new Date(c.commit.author.date),
        });
      }
    } catch {
      // skip branch on error
    }
  }

  return result;
}

function splitRepoBranch(key: string): [string, string] {
  const idx = key.indexOf(":", key.indexOf("/") + 1);
  return [key.slice(0, idx), key.slice(idx + 1)];
}

async function fetchCommitsFromSearch(
  token: string,
  login: string,
): Promise<Commit[]> {
  const commits: Commit[] = [];
  let url: string | null =
    `${GITHUB_API}/search/commits?q=${encodeURIComponent(`author:${login}`)}&sort=author-date&order=desc&per_page=100`;
  let page = 0;

  while (url && page < 3) {
    const response = await githubFetch(token, url);
    const data = (await response.json()) as GitHubSearchCommitResponse;
    commits.push(...data.items.map((item) => mapSearchCommitToCommit(item)));
    url = parseNextUrl(response.headers.get("link"));
    page++;
  }

  return commits;
}

export async function fetchPullRequests(
  token: string,
  login: string,
): Promise<PullRequest[]> {
  const items = await fetchSearchResults(
    token,
    `is:pr author:${login} sort:updated`,
  );
  return items.map(mapSearchItemToPR);
}

export async function fetchReviews(
  token: string,
  login: string,
): Promise<Review[]> {
  const items = await fetchSearchResults(
    token,
    `is:pr reviewed-by:${login} -author:${login} sort:updated`,
  );
  const reviews = await Promise.all(
    items.map(async (item) => {
      const review = mapSearchItemToReview(item);
      const prNumber = extractPullRequestNumber(item.html_url);
      if (!prNumber) return null;

      const submittedAt = await fetchLatestSubmittedReviewAt(
        token,
        review.repoName,
        prNumber,
        login,
      );
      if (!submittedAt) return null;

      return {
        ...review,
        createdAt: submittedAt,
        updatedAt: submittedAt,
      };
    }),
  );

  return reviews
    .filter((review): review is Review => review !== null)
    .toSorted((a, b) => b.updatedAt.getTime() - a.updatedAt.getTime());
}

export async function fetchCurrentPullRequests(
  token: string,
  login: string,
  repoNames: string[],
): Promise<PullRequest[]> {
  if (repoNames.length === 0) return [];

  const groups = await Promise.all(
    repoNames.map((repoName) =>
      fetchSearchResults(
        token,
        `is:pr is:open author:${login} repo:${repoName} sort:updated`,
      ),
    ),
  );

  return groups
    .flat()
    .map(mapSearchItemToPR)
    .filter(
      (pr): pr is PullRequest & { number: number } => pr.number !== undefined,
    )
    .toSorted((a, b) => b.updatedAt.getTime() - a.updatedAt.getTime());
}

export async function fetchPullRequestCommits(
  token: string,
  repoName: string,
  prNumber: number,
): Promise<Commit[]> {
  const commits: Commit[] = [];
  let url: string | null =
    `${GITHUB_API}/repos/${repoName}/pulls/${prNumber}/commits?per_page=100`;
  let page = 0;

  while (url && page < 5) {
    const response = await githubFetch(token, url);
    const data = (await response.json()) as GitHubPullRequestCommit[];

    for (const item of data) {
      const authoredAt =
        item.commit.author?.date ?? item.commit.committer?.date ?? null;
      if (!authoredAt) continue;

      commits.push({
        sha: item.sha,
        message: item.commit.message.split("\n")[0]!,
        repoName,
        branch: null,
        url: item.html_url,
        createdAt: new Date(authoredAt),
      });
    }

    url = parseNextUrl(response.headers.get("link"));
    page++;
  }

  return commits.toSorted(
    (a, b) => b.createdAt.getTime() - a.createdAt.getTime(),
  );
}

export async function fetchPullRequestReviewStatus(
  token: string,
  repoName: string,
  prNumber: number,
): Promise<PullRequestReviewStatus> {
  const reviews = await fetchPullRequestReviews(token, repoName, prNumber);
  return derivePullRequestReviewStatus(reviews);
}

export async function fetchPullRequestDiscussionComments(
  token: string,
  repoName: string,
  prNumber: number,
): Promise<PullRequestDiscussionComment[]> {
  const [issueComments, reviews, reviewComments] = await Promise.all([
    fetchIssueComments(token, repoName, prNumber),
    fetchPullRequestReviews(token, repoName, prNumber),
    fetchPullRequestReviewComments(token, repoName, prNumber),
  ]);

  const discussionComments = [
    ...issueComments
      .filter((comment) => comment.body.trim().length > 0)
      .map((comment) => ({
        id: `issue-comment:${comment.id}`,
        author: comment.user?.login ?? null,
        body: comment.body.trim(),
        url: comment.html_url,
        createdAt: new Date(comment.created_at),
        updatedAt: new Date(comment.updated_at),
      })),
    ...reviewComments
      .filter((comment) => comment.body.trim().length > 0)
      .map((comment) => ({
        id: `review-comment:${comment.id}`,
        author: comment.user?.login ?? null,
        body: comment.body.trim(),
        url: comment.html_url,
        createdAt: new Date(comment.created_at),
        updatedAt: new Date(comment.updated_at),
      })),
    ...reviews
      .filter((review) => (review.body?.trim().length ?? 0) > 0)
      .filter((review) => review.submitted_at)
      .map((review) => ({
        id: `review:${review.id ?? review.submitted_at}`,
        author: review.user?.login ?? null,
        body: review.body!.trim(),
        url:
          review.html_url ??
          `https://github.com/${repoName}/pull/${prNumber}#pullrequestreview-${review.id ?? ""}`,
        createdAt: new Date(review.submitted_at!),
        updatedAt: new Date(review.updated_at ?? review.submitted_at!),
      })),
  ];

  return discussionComments.toSorted(
    (a, b) => b.updatedAt.getTime() - a.updatedAt.getTime(),
  );
}

async function fetchPullRequestReviews(
  token: string,
  repoName: string,
  prNumber: number,
): Promise<GitHubPullRequestReview[]> {
  const reviews: GitHubPullRequestReview[] = [];
  let url: string | null =
    `${GITHUB_API}/repos/${repoName}/pulls/${prNumber}/reviews?per_page=100`;
  let page = 0;

  while (url && page < 5) {
    const response = await githubFetch(token, url);
    const data = (await response.json()) as GitHubPullRequestReview[];
    reviews.push(...data);
    url = parseNextUrl(response.headers.get("link"));
    page++;
  }

  return reviews;
}

async function fetchIssueComments(
  token: string,
  repoName: string,
  prNumber: number,
): Promise<GitHubIssueComment[]> {
  const comments: GitHubIssueComment[] = [];
  let url: string | null =
    `${GITHUB_API}/repos/${repoName}/issues/${prNumber}/comments?per_page=100`;
  let page = 0;

  while (url && page < 5) {
    const response = await githubFetch(token, url);
    const data = (await response.json()) as GitHubIssueComment[];
    comments.push(...data);
    url = parseNextUrl(response.headers.get("link"));
    page++;
  }

  return comments;
}

async function fetchPullRequestReviewComments(
  token: string,
  repoName: string,
  prNumber: number,
): Promise<GitHubPullRequestReviewComment[]> {
  const comments: GitHubPullRequestReviewComment[] = [];
  let url: string | null =
    `${GITHUB_API}/repos/${repoName}/pulls/${prNumber}/comments?per_page=100`;
  let page = 0;

  while (url && page < 5) {
    const response = await githubFetch(token, url);
    const data = (await response.json()) as GitHubPullRequestReviewComment[];
    comments.push(...data);
    url = parseNextUrl(response.headers.get("link"));
    page++;
  }

  return comments;
}

async function fetchLatestSubmittedReviewAt(
  token: string,
  repoName: string,
  prNumber: number,
  login: string,
): Promise<Date | null> {
  const reviews = await fetchPullRequestReviews(token, repoName, prNumber);
  return deriveLatestSubmittedReviewAt(reviews, login);
}

export function deriveLatestSubmittedReviewAt(
  reviews: GitHubPullRequestReview[],
  login: string,
): Date | null {
  let latestTimestamp: number | null = null;

  for (const review of reviews) {
    if (review.user?.login !== login || !review.submitted_at) continue;

    const submittedAt = new Date(review.submitted_at).getTime();
    if (!Number.isFinite(submittedAt)) continue;

    if (latestTimestamp === null || submittedAt > latestTimestamp) {
      latestTimestamp = submittedAt;
    }
  }

  return latestTimestamp === null ? null : new Date(latestTimestamp);
}

export function derivePullRequestReviewStatus(
  reviews: GitHubPullRequestReview[],
): PullRequestReviewStatus {
  const latestByReviewer = new Map<
    string,
    { state: string; submittedAt: number }
  >();

  for (const review of reviews) {
    const reviewer = review.user?.login;
    if (!reviewer || !review.submitted_at) continue;

    const normalizedState = review.state.toUpperCase();
    if (
      normalizedState !== "APPROVED" &&
      normalizedState !== "CHANGES_REQUESTED" &&
      normalizedState !== "DISMISSED"
    ) {
      continue;
    }

    const submittedAt = new Date(review.submitted_at).getTime();
    const existing = latestByReviewer.get(reviewer);
    if (!existing || submittedAt > existing.submittedAt) {
      latestByReviewer.set(reviewer, {
        state: normalizedState,
        submittedAt,
      });
    }
  }

  const effectiveStates = [...latestByReviewer.values()].map(
    (review) => review.state,
  );

  if (effectiveStates.includes("CHANGES_REQUESTED")) {
    return "changes_requested";
  }

  if (effectiveStates.includes("APPROVED")) {
    return "approved";
  }

  return "review_pending";
}

export function derivePullRequestCIStatus(
  combinedStatusState: string | null | undefined,
  combinedStatusCount: number | null | undefined,
  checkRuns: GitHubCheckRun[],
): PullRequestCIStatus {
  const normalizedCombined = combinedStatusState?.toLowerCase();
  const hasLegacyStatuses = (combinedStatusCount ?? 0) > 0;
  const normalizedChecks = checkRuns.map((checkRun) => ({
    status: checkRun.status.toLowerCase(),
    conclusion: checkRun.conclusion?.toLowerCase() ?? null,
  }));

  if (
    (normalizedCombined === "pending" && hasLegacyStatuses) ||
    normalizedChecks.some((checkRun) =>
      ["queued", "in_progress", "pending", "requested", "waiting"].includes(
        checkRun.status,
      ),
    )
  ) {
    return "ci_pending";
  }

  if (
    normalizedCombined === "failure" ||
    normalizedCombined === "error" ||
    normalizedChecks.some((checkRun) =>
      [
        "failure",
        "failed",
        "timed_out",
        "cancelled",
        "startup_failure",
        "action_required",
        "stale",
      ].includes(checkRun.conclusion ?? ""),
    )
  ) {
    return "ci_failing";
  }

  if (
    normalizedCombined === "success" ||
    normalizedChecks.some((checkRun) =>
      ["success", "neutral", "skipped"].includes(checkRun.conclusion ?? ""),
    )
  ) {
    return "ci_passing";
  }

  return "ci_unknown";
}

export async function fetchPullRequestCIStatus(
  token: string,
  repoName: string,
  prNumber: number,
): Promise<PullRequestCIStatus> {
  const prResponse = await githubFetch(
    token,
    `${GITHUB_API}/repos/${repoName}/pulls/${prNumber}`,
  );
  const pr = (await prResponse.json()) as GitHubPullRequestDetails;
  const headSha = pr.head.sha;

  const [combinedStatusResponse, checkRunsResponse] = await Promise.allSettled([
    githubFetch(
      token,
      `${GITHUB_API}/repos/${repoName}/commits/${headSha}/status`,
    ),
    githubFetch(
      token,
      `${GITHUB_API}/repos/${repoName}/commits/${headSha}/check-runs?per_page=100`,
    ),
  ]);

  const combinedStatus =
    combinedStatusResponse.status === "fulfilled"
      ? ((await combinedStatusResponse.value.json()) as GitHubCombinedStatus)
      : null;
  const checkRuns =
    checkRunsResponse.status === "fulfilled"
      ? ((await checkRunsResponse.value.json()) as GitHubCheckRunsResponse)
          .check_runs
      : [];

  return derivePullRequestCIStatus(
    combinedStatus?.state,
    combinedStatus?.total_count,
    checkRuns,
  );
}

async function fetchSearchResults(
  token: string,
  query: string,
): Promise<GitHubSearchItem[]> {
  const items: GitHubSearchItem[] = [];
  let url: string | null =
    `${GITHUB_API}/search/issues?q=${encodeURIComponent(query)}&per_page=100`;
  let page = 0;

  while (url && page < 10) {
    const response = await githubFetch(token, url);
    const data = (await response.json()) as GitHubSearchResponse;
    items.push(...data.items);
    url = parseNextUrl(response.headers.get("link"));
    page++;
  }

  return items;
}

File: internal/github/github.go
package github

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"

	"github.com/maximilianfalco/github-activity-tracker/internal/data"
)

const apiBase = "https://api.github.com"

type Client struct {
	token string
	http  *http.Client
}

type Commit struct {
	SHA       string
	Message   string
	RepoName  string
	Branch    string
	URL       string
	CreatedAt time.Time
}

type PullRequest struct {
	Title     string
	RepoName  string
	URL       string
	State     string
	CreatedAt time.Time
	UpdatedAt time.Time
}

type Review = PullRequest

type gitHubUser struct {
	Login  string `json:"login"`
	ID     int64  `json:"id"`
	Name   string `json:"name"`
	Email  string `json:"email"`
	Avatar string `json:"avatar_url"`
}

type gitHubEvent struct {
	Type      string `json:"type"`
	CreatedAt string `json:"created_at"`
	Repo      struct {
		Name string `json:"name"`
	} `json:"repo"`
	Payload struct {
		Ref string `json:"ref"`
	} `json:"payload"`
}

type repoCommit struct {
	SHA     string `json:"sha"`
	HTMLURL string `json:"html_url"`
	Commit  struct {
		Message string `json:"message"`
		Author  struct {
			Date string `json:"date"`
		} `json:"author"`
	} `json:"commit"`
}

type searchIssuesResponse struct {
	Items []struct {
		Title         string `json:"title"`
		HTMLURL       string `json:"html_url"`
		State         string `json:"state"`
		CreatedAt     string `json:"created_at"`
		UpdatedAt     string `json:"updated_at"`
		RepositoryURL string `json:"repository_url"`
		PullRequest   struct {
			MergedAt *string `json:"merged_at"`
		} `json:"pull_request"`
	} `json:"items"`
}

type searchCommitsResponse struct {
	Items []struct {
		SHA        string `json:"sha"`
		HTMLURL    string `json:"html_url"`
		Repository struct {
			FullName string `json:"full_name"`
		} `json:"repository"`
		Commit struct {
			Message string `json:"message"`
			Author  struct {
				Date string `json:"date"`
			} `json:"author"`
		} `json:"commit"`
	} `json:"items"`
}

func NewClient(token string) *Client {
	return &Client{
		token: strings.TrimSpace(token),
		http: &http.Client{
			Timeout: 20 * time.Second,
		},
	}
}

func (c *Client) FetchLogin(ctx context.Context) (string, error) {
	var user gitHubUser
	if err := c.getJSON(ctx, apiBase+"/user", "application/vnd.github.v3+json", &user); err != nil {
		return "", err
	}
	return user.Login, nil
}

func (c *Client) FetchProfile(ctx context.Context) (data.GitHubProfile, error) {
	var user gitHubUser
	if err := c.getJSON(ctx, apiBase+"/user", "application/vnd.github.v3+json", &user); err != nil {
		return data.GitHubProfile{}, err
	}
	return data.GitHubProfile{
		GitHubID: user.ID,
		Login:    user.Login,
		Name:     user.Name,
		Email:    user.Email,
		Image:    user.Avatar,
		Token:    c.token,
	}, nil
}

func (c *Client) FetchAll(ctx context.Context, login string) ([]Commit, []PullRequest, []Review, error) {
	commits, err := c.fetchCommits(ctx, login)
	if err != nil {
		return nil, nil, nil, err
	}
	prs, err := c.fetchPullRequests(ctx, login)
	if err != nil {
		return nil, nil, nil, err
	}
	reviews, err := c.fetchReviews(ctx, login)
	if err != nil {
		return nil, nil, nil, err
	}
	return commits, prs, reviews, nil
}

func (c *Client) fetchCommits(ctx context.Context, login string) ([]Commit, error) {
	eventCommits, err := c.fetchCommitsFromEvents(ctx, login)
	if err != nil {
		return nil, err
	}
	searchCommits, err := c.fetchCommitsFromSearch(ctx, login)
	if err != nil {
		return nil, err
	}

	seen := map[string]bool{}
	var merged []Commit
	for _, commit := range append(eventCommits, searchCommits...) {
		if seen[commit.SHA] {
			continue
		}
		seen[commit.SHA] = true
		merged = append(merged, commit)
	}

	sort.Slice(merged, func(i, j int) bool {
		return merged[i].CreatedAt.After(merged[j].CreatedAt)
	})
	return merged, nil
}

func (c *Client) fetchCommitsFromEvents(ctx context.Context, login string) ([]Commit, error) {
	var events []gitHubEvent
	if err := c.getJSON(ctx, fmt.Sprintf("%s/users/%s/events?per_page=100", apiBase, login), "application/vnd.github.v3+json", &events); err != nil {
		return nil, err
	}

	branches := map[string]bool{}
	for _, event := range events {
		if event.Type != "PushEvent" || event.Payload.Ref == "" {
			continue
		}
		branch := strings.TrimPrefix(event.Payload.Ref, "refs/heads/")
		branches[event.Repo.Name+":"+branch] = true
	}

	baseBranches := map[string]bool{"main": true, "master": true, "next": true, "develop": true, "dev": true}
	branchKeys := make([]string, 0, len(branches))
	for key := range branches {
		branchKeys = append(branchKeys, key)
	}
	sort.Slice(branchKeys, func(i, j int) bool {
		_, branchA := splitRepoBranch(branchKeys[i])
		_, branchB := splitRepoBranch(branchKeys[j])
		aBase := 1
		bBase := 1
		if baseBranches[branchA] {
			aBase = 0
		}
		if baseBranches[branchB] {
			bBase = 0
		}
		return aBase < bBase
	})

	since := time.Now().Add(-90 * 24 * time.Hour).UTC().Format(time.RFC3339)
	seen := map[string]bool{}
	var result []Commit
	for _, key := range branchKeys {
		repoName, branch := splitRepoBranch(key)
		endpoint := fmt.Sprintf("%s/repos/%s/commits?author=%s&sha=%s&per_page=100&since=%s", apiBase, repoName, url.QueryEscape(login), url.QueryEscape(branch), url.QueryEscape(since))
		var items []repoCommit
		if err := c.getJSON(ctx, endpoint, "application/vnd.github.v3+json", &items); err != nil {
			continue
		}
		for _, item := range items {
			if seen[item.SHA] {
				continue
			}
			seen[item.SHA] = true
			createdAt, _ := time.Parse(time.RFC3339, item.Commit.Author.Date)
			result = append(result, Commit{
				SHA:       item.SHA,
				Message:   firstLine(item.Commit.Message),
				RepoName:  repoName,
				Branch:    branch,
				URL:       item.HTMLURL,
				CreatedAt: createdAt,
			})
		}
	}

	return result, nil
}

func (c *Client) fetchCommitsFromSearch(ctx context.Context, login string) ([]Commit, error) {
	query := url.QueryEscape("author:" + login)
	endpoint := fmt.Sprintf("%s/search/commits?q=%s&sort=author-date&order=desc&per_page=100", apiBase, query)
	var response searchCommitsResponse
	if err := c.getJSON(ctx, endpoint, "application/vnd.github.cloak-preview+json", &response); err != nil {
		return nil, err
	}

	result := make([]Commit, 0, len(response.Items))
	for _, item := range response.Items {
		createdAt, _ := time.Parse(time.RFC3339, item.Commit.Author.Date)
		result = append(result, Commit{
			SHA:       item.SHA,
			Message:   firstLine(item.Commit.Message),
			RepoName:  item.Repository.FullName,
			URL:       item.HTMLURL,
			CreatedAt: createdAt,
		})
	}

	return result, nil
}

func (c *Client) fetchPullRequests(ctx context.Context, login string) ([]PullRequest, error) {
	items, err := c.fetchIssues(ctx, "is:pr author:"+login+" sort:updated")
	if err != nil {
		return nil, err
	}
	return items, nil
}

func (c *Client) fetchReviews(ctx context.Context, login string) ([]Review, error) {
	items, err := c.fetchIssues(ctx, "is:pr reviewed-by:"+login+" -author:"+login+" sort:updated")
	if err != nil {
		return nil, err
	}
	return items, nil
}

func (c *Client) fetchIssues(ctx context.Context, query string) ([]PullRequest, error) {
	endpoint := fmt.Sprintf("%s/search/issues?q=%s&per_page=100", apiBase, url.QueryEscape(query))
	var response searchIssuesResponse
	if err := c.getJSON(ctx, endpoint, "application/vnd.github.v3+json", &response); err != nil {
		return nil, err
	}

	result := make([]PullRequest, 0, len(response.Items))
	for _, item := range response.Items {
		createdAt, _ := time.Parse(time.RFC3339, item.CreatedAt)
		updatedAt, _ := time.Parse(time.RFC3339, item.UpdatedAt)
		state := "closed"
		if item.PullRequest.MergedAt != nil {
			state = "merged"
		} else if item.State == "open" {
			state = "open"
		}
		result = append(result, PullRequest{
			Title:     item.Title,
			RepoName:  extractRepoName(item.RepositoryURL),
			URL:       item.HTMLURL,
			State:     state,
			CreatedAt: createdAt,
			UpdatedAt: updatedAt,
		})
	}

	return result, nil
}

func (c *Client) getJSON(ctx context.Context, endpoint, accept string, target any) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return fmt.Errorf("build github request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+c.token)
	req.Header.Set("Accept", accept)

	resp, err := c.http.Do(req)
	if err != nil {
		return fmt.Errorf("call github: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusForbidden && resp.Header.Get("X-RateLimit-Remaining") == "0" {
		return errors.New("GitHub rate limit exceeded")
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("GitHub API error: %s", resp.Status)
	}
	return json.NewDecoder(resp.Body).Decode(target)
}

func splitRepoBranch(value string) (string, string) {
	parts := strings.SplitN(value, ":", 2)
	if len(parts) != 2 {
		return value, ""
	}
	return parts[0], parts[1]
}

func extractRepoName(repositoryURL string) string {
	parts := strings.Split(repositoryURL, "/repos/")
	if len(parts) == 2 {
		return parts[1]
	}
	return repositoryURL
}

func firstLine(message string) string {
	parts := strings.Split(message, "\n")
	return parts[0]
}

func BuildCacheEntries(userID string, commits []Commit, prs []PullRequest, reviews []Review) []data.Activity {
	fetchedAt := time.Now().UTC()
	seen := map[string]bool{}
	var result []data.Activity

	for _, commit := range commits {
		if seen[commit.SHA] {
			continue
		}
		seen[commit.SHA] = true
		result = append(result, data.Activity{
			ID:        "ghat_commit_" + commit.SHA,
			Type:      "commit",
			RepoName:  commit.RepoName,
			Title:     commit.Message,
			URL:       commit.URL,
			SHA:       validString(commit.SHA),
			Branch:    validString(commit.Branch),
			CreatedAt: commit.CreatedAt,
			FetchedAt: fetchedAt,
		})
	}
	for _, pr := range prs {
		result = append(result, data.Activity{
			ID:        activityID("pr", pr.URL),
			Type:      "pr",
			RepoName:  pr.RepoName,
			Title:     pr.Title,
			URL:       pr.URL,
			State:     validString(pr.State),
			CreatedAt: pr.CreatedAt,
			UpdatedAt: validTime(pr.UpdatedAt),
			FetchedAt: fetchedAt,
		})
	}
	for _, review := range reviews {
		result = append(result, data.Activity{
			ID:        activityID("review", review.URL),
			Type:      "review",
			RepoName:  review.RepoName,
			Title:     review.Title,
			URL:       review.URL,
			State:     validString(review.State),
			CreatedAt: review.CreatedAt,
			UpdatedAt: validTime(review.UpdatedAt),
			FetchedAt: fetchedAt,
		})
	}
	return result
}

func activityID(prefix, value string) string {
	replacer := strings.NewReplacer("https://", "", "http://", "", "/", "_", ":", "_", "?", "_", "&", "_", "=", "_", "-", "_")
	return "ghat_" + prefix + "_" + replacer.Replace(value)
}

func validString(value string) sql.NullString {
	return sql.NullString{String: value, Valid: value != ""}
}

func validTime(value time.Time) sql.NullTime {
	return sql.NullTime{Time: value, Valid: !value.IsZero()}
}
